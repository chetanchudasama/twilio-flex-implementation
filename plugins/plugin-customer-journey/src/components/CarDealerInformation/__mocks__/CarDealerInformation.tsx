import React from "react";
import { Props } from "../CarDealerInformation";

const CarDealerInformation: React.FC<Props> = (props) => <div />;

export default CarDealerInformation;
