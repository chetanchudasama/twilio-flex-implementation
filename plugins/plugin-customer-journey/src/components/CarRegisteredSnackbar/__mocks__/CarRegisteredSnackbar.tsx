import React from "react";
import { CarRegisteredSnackbarProps } from "../CarRegisteredSnackbar.Props";

export const CarRegisteredSnackbar: React.FC<CarRegisteredSnackbarProps> = (
  props
) => {
  return <div className="car-registered-snackbar">CarRegisteredSnackbar</div>;
};
