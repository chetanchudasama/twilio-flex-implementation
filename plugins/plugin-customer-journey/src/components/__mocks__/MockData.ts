import { WorkerAttributes } from "@twilio/flex-ui";
import { namespace } from "../../states";

const workerAttributes: WorkerAttributes = {
  contact_uri: "client:test",
  email: "test@test.com",
  full_name: "test user",
  image_url: "img_url",
  roles: ["admin"],
  agentId: 1,
};

const conferenceSource = {
  source: {
    participants: [
      {
        callSid: "callSid",
        participantType: "customer",
        workerSid: "",
      },
      {
        callSid: "callSid",
        participantType: "worker",
        workerSid: "workerSid",
      },
    ],
  },
};

const conferenceState = new Map();
conferenceState.set("taskSid", conferenceSource);

export const flexManagerMock = jest.fn().mockReturnValue({
  workerClient: {
    source: {},
    sid: "workerSid",
    name: "test user",
    fullName: "test user",
    activityName: "Offline",
    attributes: workerAttributes,
    dateUpdated: new Date(),
    isAvailable: false,
    activityDuration: "",
    dateStatusChanged: new Date(1).toISOString(),
    activity: {
      available: false,
      name: "Offline",
      sid: "OfflineSid",
    },
  },
  store: {
    getState: jest.fn().mockReturnValue({
      flex: {
        worker: {
          attributes: workerAttributes,
        },
        session: {
          ssoTokenPayload: {
            token: "ssoToken",
          },
        },
        conferences: {
          states: conferenceState,
        },
      },
      [namespace]: {
        phoneNumberList: {
          phoneNumberList: [
            {
              friendlyName: "Sales",
              phoneNumber: "+1234567891",
              queueSid: "QUEUE_SID",
            },
          ],
        },
      },
    }),
  },
  strings: {
    DIALPADExternalTransferSearchInputLabel: "Search By Name or PhoneNumber",
    DIALPADExternalTransferHoverOver: "Add",
    DIALPADExternalTransferPopupHeader: "External Transfer",
  },
  serviceConfiguration: {
    outbound_call_flows: {
      default: {
        caller_id: "CALLER_ID",
      },
    },
  },
});
