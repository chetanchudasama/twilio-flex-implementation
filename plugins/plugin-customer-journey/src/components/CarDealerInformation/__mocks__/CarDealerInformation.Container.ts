import CarDealerInformation from "./CarDealerInformation";

export default CarDealerInformation;
