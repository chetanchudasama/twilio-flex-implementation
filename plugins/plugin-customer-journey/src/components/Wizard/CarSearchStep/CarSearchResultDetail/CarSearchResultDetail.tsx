import { OfferInformationModel } from "models/OfferInformationModel";
import React, { useMemo, useState } from "react";

import { secondary, Shared, WhatsAppIcon } from "@common/components";
import { Button } from "@material-ui/core";
import Grid from "@material-ui/core/Grid";
import ChatBubbleOutlineIcon from "@material-ui/icons/ChatBubbleOutline";
import FavoriteIcon from "@material-ui/icons/Favorite";
import FavoriteBorderRoundedIcon from "@material-ui/icons/FavoriteBorderRounded";

import { ChannelType, DealerStateType } from "../../../../common/enum";
import { QuoteDetailModel } from "../../../../models/QuoteDetailModel";
import { VehicleDetailModel } from "../../../../models/VehicleDetailModel";
import { VehicleSummaryDetailModel } from "../../../../models/VehicleSummaryDetailModel";
import CarDealerInformation from "../../../CarDealerInformation/CarDealerInformation.Container";
import CarDetailViewDialog from "../../../CarDetailViewDialog/CarDetailViewDialog.Container";
import { VehicleImage } from "../../../VehicleImage/VehicleImage";
import QuoteDialog from "../../QuotesStep/QuoteDialog/QuoteDialog.Container";
import { CarSearchResultDetailStyles } from "./CarSearchResultDetail.Styles";

export interface CarSearchResultDetailProp {
  vehicleDetail: VehicleSummaryDetailModel | VehicleDetailModel;
  isCarSaved: boolean;
  sendCarDetail: (channelType: ChannelType) => void;
  saveCarDetail: () => void;
}

const CarSearchResultDetail: React.FC<CarSearchResultDetailProp> = ({
  vehicleDetail,
  isCarSaved,
  sendCarDetail,
  saveCarDetail,
}) => {
  const getCarMakeModel = () => {
    return Shared.getFormattedText(
      [vehicleDetail.make, vehicleDetail.model],
      " | "
    );
  };
  const [selectedOfferTerm, setSelectedOfferTerm] = useState(36);
  const [openCarDetailDialog, setOpenCarDetailDialog] =
    useState<boolean>(false);
  const [openCreateQuoteDialog, setOpenCreateQuoteDialog] = useState(false);
  const [quoteDetail, setQuoteDetail] = useState(new QuoteDetailModel());

  // TODO: remove once fetched from API
  const offerInformation: OfferInformationModel[] = [
    {
      term: 36,
      monthlyAmount: 2000,
      bestMatch: true,
      available: true,
    },
    {
      term: 48,
      monthlyAmount: 5000,
      bestMatch: false,
      available: false,
    },
    {
      term: 60,
      monthlyAmount: 8000,
      bestMatch: false,
      available: true,
    },
  ];

  const getOfferClassName = (offer: OfferInformationModel) => {
    if (offer.term === selectedOfferTerm) {
      return "offer-selected";
    }

    if (offer.bestMatch) {
      return "offer-best-match";
    }

    if (!offer.available) {
      return "offer-unavailable";
    }

    return "offer-detail";
  };

  const openQuoteDialog = () => {
    // assign vehicle detail to quote detail
    const newQuoteDetail = new QuoteDetailModel(vehicleDetail);
    newQuoteDetail.term =
      Shared.termList.find((t) => t.id === selectedOfferTerm)?.id ||
      Shared.termList[0].id;
    setQuoteDetail(newQuoteDetail);
    // open quote dialog
    setOpenCreateQuoteDialog(true);
  };

  const image = useMemo(() => {
    if (vehicleDetail instanceof VehicleSummaryDetailModel) {
      return vehicleDetail.image;
    }
    return vehicleDetail.images.length > 0 ? vehicleDetail.images[0] : "";
  }, [vehicleDetail]);

  const imagesCount = useMemo(() => {
    if (vehicleDetail instanceof VehicleSummaryDetailModel) {
      return vehicleDetail.imagesCount;
    }
    return vehicleDetail.images.length;
  }, [vehicleDetail]);

  const dealerAdminFee = useMemo(() => {
    if (vehicleDetail instanceof VehicleSummaryDetailModel) {
      return vehicleDetail.dealerAdminFee || 0;
    }
    return 0;
  }, [vehicleDetail]);

  const offerTermFunction = (offer: OfferInformationModel) => {
    if (offer.available) {
      setSelectedOfferTerm(offer.term);
    }
  };

  const disableSendMessageButton = useMemo(() => {
    return (
      !vehicleDetail.is247Cars ||
      vehicleDetail.dealer.state !== DealerStateType.Approved
    );
  }, [vehicleDetail.dealer.state, vehicleDetail.is247Cars]);

  return (
    <>
      <CarSearchResultDetailStyles>
        <Grid container>
          <Grid item xs={4} className="vehicle-img-container">
            <VehicleImage
              image={image}
              imageList={vehicleDetail.images}
              imagesCount={imagesCount}
              showImageCount
            />
          </Grid>
          <Grid item xs={8}>
            <Grid container className="car-search-detail">
              <Grid item xs={8}>
                <Grid container spacing={8} className="car-carousel">
                  <Grid item xs={12}>
                    <div className="car-price">
                      {Shared.getFormattedCurrencyValue(vehicleDetail.price)}
                    </div>
                  </Grid>
                  <Grid item xs={12}>
                    <div className="car-model">{getCarMakeModel()}</div>
                  </Grid>
                  <Grid item xs={12}>
                    <ul className="car-specification">
                      <li>
                        {`${Shared.getFormattedNumber(
                          vehicleDetail.mileage
                        )} miles`}
                      </li>
                      <li>{vehicleDetail.regNumber}</li>
                      <li>{vehicleDetail.regYear}</li>
                      <li>{vehicleDetail.trans}</li>
                      <li>{vehicleDetail.fuel}</li>
                    </ul>
                  </Grid>
                  <Grid item xs={12} className="view-more-detail">
                    <Button
                      className="view-more-detail-link"
                      onClick={() => setOpenCarDetailDialog(true)}
                      disableRipple
                    >
                      View More Detail
                    </Button>
                  </Grid>
                  <Grid item xs={12}>
                    <Button
                      variant="contained"
                      onClick={openQuoteDialog}
                      className="quote-btn"
                    >
                      Quote
                    </Button>
                    <Button
                      variant="outlined"
                      className="secondary-button"
                      onClick={saveCarDetail}
                    >
                      {isCarSaved ? (
                        <FavoriteIcon className="icon" />
                      ) : (
                        <FavoriteBorderRoundedIcon className="icon" />
                      )}
                    </Button>
                    <Button
                      variant="outlined"
                      className="secondary-button whatsapp-icon-button"
                      onClick={() => sendCarDetail(ChannelType.Whatsapp)}
                      disabled={disableSendMessageButton}
                    >
                      <WhatsAppIcon
                        color={!disableSendMessageButton ? secondary : ""}
                        opacity={disableSendMessageButton ? "0.26" : ""}
                      />
                    </Button>
                    <Button
                      variant="outlined"
                      className="secondary-button sms-icon-button"
                      onClick={() => sendCarDetail(ChannelType.SMS)}
                      disabled={disableSendMessageButton}
                    >
                      <ChatBubbleOutlineIcon className="icon" />
                    </Button>
                  </Grid>
                </Grid>
              </Grid>
              <Grid item xs={4}>
                {offerInformation.map((offer) => (
                  <React.Fragment key={offer.term}>
                    <div
                      className={getOfferClassName(offer)}
                      onClick={() => offerTermFunction(offer)}
                      onKeyDown={() => offerTermFunction(offer)}
                      role="button"
                      tabIndex={0}
                    >
                      {offer.bestMatch && (
                        <div
                          className={
                            offer.term === selectedOfferTerm
                              ? "offer-best-match-selected"
                              : "offer-best-match-heading"
                          }
                        >
                          Best Match
                        </div>
                      )}
                      <div className="offer-detail-content">
                        <div className="offer-term">
                          {offer.term} months from
                        </div>
                        <div className="offer-price">
                          {!offer.available ? (
                            <div className="offer-amount">Unavailable</div>
                          ) : (
                            <>
                              <div className="offer-amount">
                                {Shared.getFormattedCurrencyValue(
                                  offer.monthlyAmount
                                )}
                              </div>
                              <div className="offer-pm">/pm</div>
                            </>
                          )}
                        </div>
                      </div>
                    </div>
                  </React.Fragment>
                ))}
                <div className="dealer-fee">
                  {`Excl. ${Shared.getFormattedCurrencyValue(
                    dealerAdminFee
                  )} dealer fee`}
                </div>
              </Grid>
            </Grid>
            {/* <Grid container className="dealer-detail">
              <Grid item xs={12}>
                <CarDealerInformation
                  dealerInfo={vehicleDetail.dealer}
                  distance={vehicleDetail.distance!}
                />
              </Grid>
            </Grid> */}
          </Grid>
        </Grid>
      </CarSearchResultDetailStyles>
      {openCarDetailDialog && (
        <CarDetailViewDialog
          open={openCarDetailDialog}
          vehicleId={vehicleDetail.vehicleId}
          handleCloseDialog={() => setOpenCarDetailDialog(false)}
        />
      )}
      {openCreateQuoteDialog && (
        <QuoteDialog
          open={openCreateQuoteDialog}
          quoteDetail={quoteDetail}
          handleCloseDialog={() => setOpenCreateQuoteDialog(false)}
        />
      )}
    </>
  );
};

export default CarSearchResultDetail;
