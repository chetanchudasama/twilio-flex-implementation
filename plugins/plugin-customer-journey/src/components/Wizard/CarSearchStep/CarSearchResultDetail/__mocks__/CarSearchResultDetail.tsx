import React from "react";

import { CarSearchResultDetailProp } from "../CarSearchResultDetail";

const CarSearchResultDetail: React.FC<CarSearchResultDetailProp> = (props) => (
  <div />
);

export default CarSearchResultDetail;
