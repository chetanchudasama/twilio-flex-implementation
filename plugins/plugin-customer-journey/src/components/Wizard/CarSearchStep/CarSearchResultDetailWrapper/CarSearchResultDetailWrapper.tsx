import { SnackbarMessage } from "@common/components";
import React, { useMemo, useState } from "react";
import * as Flex from "@twilio/flex-ui";
import { Manager } from "@twilio/flex-ui";
import { AppState } from "../../../../states";
import { useApplicationService } from "../../../../services/application.service";
import { ChannelType } from "../../../../common/enum";
import { showErrorMessage } from "../../../../Notifications";
import { SendCarRequestModel } from "../../../../models/SendCarRequestModel";
import { VehicleSummaryDetailModel } from "../../../../models/VehicleSummaryDetailModel";
import { VehicleDetailModel } from "../../../../models/VehicleDetailModel";
import CarSearchResultDetail from "../CarSearchResultDetail/CarSearchResultDetail";
import {
  StateToProps,
  DispatchToProps,
} from "./CarSearchResultDetailWrapper.Props";

import { useVehicleService } from "../../../../services/vehicle.service";

export interface CarSearchResultDetailWrapperProp {
  vehicleDetail: VehicleSummaryDetailModel | VehicleDetailModel;
}

export type ComponentProps = StateToProps &
  DispatchToProps &
  CarSearchResultDetailWrapperProp;

const CarSearchResultDetailWrapper: React.FC<ComponentProps> = ({
  applicationId,
  vehicleDetail,
  savedVehicles,
  mobileNumber,
  postCode,
  setSavedVehicles,
}) => {
  const state: AppState = Manager.getInstance().store.getState();
  const applicationService = useApplicationService(
    state.flex.session.ssoTokenPayload.token
  );

  const phoenixToken: string = useMemo(() => {
    return state.crm?.crmState?.phoenixToken ?? "";
  }, [state.crm?.crmState?.phoenixToken]);

  const vehicleService = useVehicleService(
    state.flex.session.ssoTokenPayload.token,
    phoenixToken
  );
  const [showSnackbar, setShowSnackbar] = useState(false);
  const [snackbarMessage, setSnackbarMessage] = useState("");
  const [snackbarButtonText, setSnackbarButtonText] = useState("");

  const resetSnackbarDetail = () => {
    setShowSnackbar(false);
    setSnackbarMessage("");
    setSnackbarButtonText("");
  };

  const handleSendCarDetail = async (channelType: ChannelType) => {
    resetSnackbarDetail();
    try {
      const sendCarDetailModel: SendCarRequestModel = new SendCarRequestModel();
      sendCarDetailModel.applicationId = applicationId;
      sendCarDetailModel.smsTypeId = channelType;
      sendCarDetailModel.mobileNumber = mobileNumber;
      sendCarDetailModel.vehicleSearchUrl = vehicleDetail.dealer.url;

      await applicationService.sendCarDetail(sendCarDetailModel);
      setSnackbarMessage(
        `Car detail sent via ${ChannelType[channelType]} successfully!`
      );
      setShowSnackbar(true);
    } catch (error) {
      showErrorMessage("Something went wrong, please try again!", "", true);
    }
  };

  const handleSaveCarDetail = async () => {
    resetSnackbarDetail();

    try {
      const carDetail = await vehicleService.getVehicleDetail(
        vehicleDetail.vehicleId,
        postCode
      );

      const isCarAlreadySaved = savedVehicles.some(
        (v) => v.vehicleId === vehicleDetail.vehicleId
      );

      let newSavedVehicles = [...savedVehicles];

      if (isCarAlreadySaved) {
        newSavedVehicles = savedVehicles.filter(
          (v) => v.vehicleId !== vehicleDetail.vehicleId
        );
      } else {
        newSavedVehicles.push(carDetail);
      }

      await applicationService.updateSavedVehicles(
        applicationId,
        newSavedVehicles
      );

      // show snackbar message
      if (isCarAlreadySaved) {
        setSnackbarMessage("This car has been removed from saved cars.");
        setSnackbarButtonText("View Saved Cars");
      } else {
        setSnackbarMessage("This car has been added to saved cars.");
        setSnackbarButtonText("View Saved Cars");
      }
      setShowSnackbar(true);

      setSavedVehicles(newSavedVehicles);
    } catch (error) {
      showErrorMessage("Something went wrong, please try again!", "", true);
    }
  };

  const handleCloseSnackbar = (isButtonClicked: boolean) => {
    if (snackbarButtonText && isButtonClicked) {
      Flex.Actions.invokeAction("MoveToCarSearchStep", { tabIndex: 1 });
    }
    resetSnackbarDetail();
  };

  return (
    <>
      <CarSearchResultDetail
        vehicleDetail={vehicleDetail}
        sendCarDetail={handleSendCarDetail}
        saveCarDetail={handleSaveCarDetail}
        isCarSaved={savedVehicles.some(
          (vehicle: VehicleDetailModel) =>
            vehicle.vehicleId === vehicleDetail.vehicleId
        )}
      />

      {showSnackbar && (
        <SnackbarMessage
          open={showSnackbar}
          message={snackbarMessage}
          buttonText={snackbarButtonText}
          handleCloseSnackbar={handleCloseSnackbar}
        />
      )}
    </>
  );
};

export default CarSearchResultDetailWrapper;
