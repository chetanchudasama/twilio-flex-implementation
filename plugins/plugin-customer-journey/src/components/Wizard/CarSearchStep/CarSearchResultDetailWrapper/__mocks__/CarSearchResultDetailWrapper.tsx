import React from "react";
import { ComponentProps } from "../CarSearchResultDetailWrapper";

const CarSearchResultDetailWrapper: React.FC<ComponentProps> = (props) => (
  <div />
);

export default CarSearchResultDetailWrapper;
