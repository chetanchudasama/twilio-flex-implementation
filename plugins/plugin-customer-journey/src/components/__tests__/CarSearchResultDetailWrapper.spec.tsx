import { Manager } from "@twilio/flex-ui";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import flushPromises from "flush-promises";
import React from "react";
import { create } from "react-test-renderer";
import { SnackbarMessage } from "@common/components";
import { SendCarRequestModel } from "../../models/SendCarRequestModel";
import CarSearchResultDetailWrapper, {
  CarSearchResultDetailWrapperProp,
} from "../Wizard/CarSearchStep/CarSearchResultDetailWrapper/CarSearchResultDetailWrapper";
import {
  DispatchToProps,
  StateToProps,
} from "../Wizard/CarSearchStep/CarSearchResultDetailWrapper/CarSearchResultDetailWrapper.Props";
import { VehicleSummaryDetailModel } from "../../models/VehicleSummaryDetailModel";
import CarSearchResultDetail from "../Wizard/CarSearchStep/CarSearchResultDetail/CarSearchResultDetail";
import { showErrorMessage } from "../../Notifications";
import { ChannelType } from "../../common/enum";
import { flexManagerMock } from "../__mocks__/MockData";
import { VehicleDetailModel } from "../../models/VehicleDetailModel";

const vehicleDetail = new VehicleSummaryDetailModel();
vehicleDetail.vehicleId = 1;

const savedCar = new VehicleDetailModel();
savedCar.vehicleId = 1;

const sendCarRequestModel = new SendCarRequestModel();
sendCarRequestModel.applicationId = 1;
sendCarRequestModel.smsTypeId = 1;

const core: StateToProps & DispatchToProps & CarSearchResultDetailWrapperProp =
  {
    vehicleDetail,
    applicationId: 1,
    savedVehicles: [],
    mobileNumber: "",
    postCode: "",
    setSavedVehicles: jest.fn(),
  };
let props: StateToProps & DispatchToProps & CarSearchResultDetailWrapperProp = {
  ...core,
};

jest.mock("../../Notifications", () => {
  return {
    showErrorMessage: jest.fn(() => {}),
  };
});

jest.mock("../../services/application.service");
jest.mock("../../services/vehicle.service");

const mockSendCarDetail = jest.fn();
const mockDeleteSavedCar = jest.fn();
const mockGetVehicleDetail = jest.fn();
const mockUpdateSavedVehicles = jest.fn();

jest.mock("../../services/application.service", () => ({
  useApplicationService: jest.fn(() => ({
    sendCarDetail: mockSendCarDetail,
    deleteSavedCar: mockDeleteSavedCar,
    updateSavedVehicles: mockUpdateSavedVehicles,
  })),
}));
jest.mock("../../services/vehicle.service", () => ({
  useVehicleService: jest.fn(() => ({
    getVehicleDetail: mockGetVehicleDetail,
  })),
}));

jest.mock(
  "../Wizard/CarSearchStep/CarSearchResultDetail/CarSearchResultDetail"
);

Manager.getInstance = flexManagerMock;

describe("CarSearchResultDetailWrapper", () => {
  beforeEach(() => {
    props = { ...core };
  });
  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();
  });

  it("renders correctly", () => {
    const wrapper = create(
      <CarSearchResultDetailWrapper {...props} />
    ).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it("when saveCarDetail is called from CarSearchResultDetail component, vehicleDetail should be fetched for vehicle", async () => {
    mockGetVehicleDetail.mockImplementationOnce(() =>
      Promise.resolve(savedCar)
    );
    mockUpdateSavedVehicles.mockImplementationOnce(() => Promise.resolve());

    const wrapper = mount(<CarSearchResultDetailWrapper {...props} />);

    expect(wrapper.find(CarSearchResultDetail).exists()).toBeTruthy();

    await act(async () => {
      wrapper
        .find(CarSearchResultDetail)
        .invoke<"saveCarDetail">("saveCarDetail")();
      wrapper.update();
      await flushPromises();
      wrapper.update();

      expect(mockGetVehicleDetail).toHaveBeenCalled();
    });
  });

  it("should show error message and should not update savedVehicles list, if api for fetching vehicle gives error", async () => {
    mockGetVehicleDetail.mockImplementationOnce(() =>
      Promise.reject({ statusCode: 500 })
    );

    const wrapper = mount(<CarSearchResultDetailWrapper {...props} />);

    expect(wrapper.find(CarSearchResultDetail).exists()).toBeTruthy();

    await act(async () => {
      wrapper
        .find(CarSearchResultDetail)
        .invoke<"saveCarDetail">("saveCarDetail")();
      wrapper.update();
      await flushPromises();
      wrapper.update();
    });

    expect(showErrorMessage).toHaveBeenCalledWith(
      "Something went wrong, please try again!",
      "",
      true
    );
    expect(mockUpdateSavedVehicles).not.toHaveBeenCalled();
  });

  it("when saveCarDetail is called from CarSearchResultDetail component, saved vehicleList should be updated", async () => {
    mockGetVehicleDetail.mockImplementationOnce(() =>
      Promise.resolve(savedCar)
    );
    mockUpdateSavedVehicles.mockImplementationOnce(() =>
      Promise.resolve(undefined)
    );

    const wrapper = mount(<CarSearchResultDetailWrapper {...props} />);

    expect(wrapper.find(CarSearchResultDetail).exists()).toBeTruthy();

    await act(async () => {
      wrapper
        .find(CarSearchResultDetail)
        .invoke<"saveCarDetail">("saveCarDetail")();
      wrapper.update();
      await flushPromises();
      wrapper.update();

      expect(mockUpdateSavedVehicles).toHaveBeenCalledWith(
        props.applicationId,
        [savedCar]
      );
      expect(props.setSavedVehicles).toHaveBeenCalledWith([savedCar]);
    });
  });

  it("show error message and should not update stored list of saved vehicles, if api for updating saved vehicleList returns error", async () => {
    mockGetVehicleDetail.mockImplementationOnce(() =>
      Promise.resolve(savedCar)
    );
    mockUpdateSavedVehicles.mockImplementationOnce(() =>
      Promise.reject({ statusCode: 500 })
    );

    const wrapper = mount(<CarSearchResultDetailWrapper {...props} />);

    expect(wrapper.find(CarSearchResultDetail).exists()).toBeTruthy();

    await act(async () => {
      wrapper
        .find(CarSearchResultDetail)
        .invoke<"saveCarDetail">("saveCarDetail")();
      wrapper.update();
      await flushPromises();
      wrapper.update();

      expect(SnackbarMessage).toBeTruthy();
      expect(showErrorMessage).toHaveBeenCalledWith(
        "Something went wrong, please try again!",
        "",
        true
      );
      expect(props.setSavedVehicles).not.toHaveBeenCalled();
    });
  });

  it("when sendCarDetail is called from CarSearchResultDetail component, sendCarDetail API should be called", async () => {
    mockSendCarDetail.mockImplementationOnce(() => Promise.resolve(undefined));

    const wrapper = mount(<CarSearchResultDetailWrapper {...props} />);

    expect(wrapper.find(CarSearchResultDetail).exists()).toBeTruthy();

    await act(async () => {
      wrapper
        .find(CarSearchResultDetail)
        .invoke<"sendCarDetail">("sendCarDetail")(ChannelType.SMS);
      await flushPromises();

      expect(mockSendCarDetail).toHaveBeenCalledWith(sendCarRequestModel);
      expect(SnackbarMessage).toBeTruthy();
    });
  });

  it("when sendCarDetail API returns an error, error message should be displayed", async () => {
    mockSendCarDetail.mockImplementationOnce(() =>
      Promise.reject({ statusCode: 500 })
    );

    const wrapper = mount(<CarSearchResultDetailWrapper {...props} />);

    expect(wrapper.find(CarSearchResultDetail).exists()).toBeTruthy();

    await act(async () => {
      wrapper
        .find(CarSearchResultDetail)
        .invoke<"sendCarDetail">("sendCarDetail")(ChannelType.SMS);
      await flushPromises();

      expect(mockSendCarDetail).toHaveBeenCalledWith(sendCarRequestModel);
      expect(showErrorMessage).toHaveBeenCalledWith(
        "Something went wrong, please try again!",
        "",
        true
      );
    });
  });
});
