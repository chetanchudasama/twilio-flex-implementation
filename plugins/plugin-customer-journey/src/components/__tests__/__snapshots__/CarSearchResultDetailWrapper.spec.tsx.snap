// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarSearchResultDetailWrapper renders correctly 1`] = `<div />`;
